### PR TITLE
Add state manager

### PR DESCRIPTION
**Pull request by oleksandr-didyk**
Resolves https://github.com/dotnet/arcade-services/issues/3805

Adds a State Manager and its factory for managing Redis connection

### CODE REVIEW

File: src/ProductConstructionService/ProductConstructionService.Api/PcsStartup.cs [hunk 1]
```diff
@@ -124,12 +124,12 @@ static PcsStartup()
...
     /// <param name="builder"></param>
     /// <param name="addKeyVault">Use KeyVault for secrets?</param>
-    /// <param name="addRedis">Use Redis for caching?</param>
+    /// <param name="addRedisAuth">Use authenticated connection for Redis?</param>
     /// <param name="addSwagger">Add Swagger UI?</param>
     internal static async Task ConfigurePcs(
         this WebApplicationBuilder builder,
         bool addKeyVault,
-        bool addRedis,
+        bool addRedisAuth,
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Api/PcsStartup.cs:132]: `authRedis` might be better - this sounds like we're addng some auth based on / stored in Redis

File: src/ProductConstructionService/ProductConstructionService.Common/StateManager.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    internal StateManager(IConnectionMultiplexer connection, string stateKey)
+    {
+        _connection = connection;
+        _stateKey = stateKey;
+    }
+
+    public void Set(string value)
+    {
+        var cache = _connection.GetDatabase();
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Common/StateManager.cs:29]: Maybe we can just have `private TSomethingHere Cache => _connection.GetDatabase()`?

File: src/ProductConstructionService/ProductConstructionService.Common/StateManagerFactory.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+namespace ProductConstructionService.Common;
+
+public interface IStateManagerFactory
+{
+    public IStateManager CreateCodeFlowStateManager(string actorId);
+    public IStateManager CreateInProgressPrStateManager(string actorId);
+    public IStateManager CreateUpdateAssetsStateManager(string actorId);
+}
+
+public class StateManagerFactory: IStateManagerFactory
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Common/StateManagerFactory.cs:15]: This is nice. I will move it to the DependencyFlow project, once it's added

**Review by premun — COMMENTED** [review]
Just nits

File: src/ProductConstructionService/ProductConstructionService.Api/PcsStartup.cs (hunk 1 above)
Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.Api/PcsStartup.cs:132]: Good point, thanks. Will change

File: src/ProductConstructionService/ProductConstructionService.Common/StateManager.cs (hunk 2 above)
Comment (oleksandr-didyk) [reply src/ProductConstructionService/ProductConstructionService.Common/StateManager.cs:29]: The docs state that this object is a throwaway - https://stackexchange.github.io/StackExchange.Redis/Basics#using-a-redis-database

Based on this I had it created for every operation. Don't have a strong opinion though, we can just store it

Comment (premun) [reply src/ProductConstructionService/ProductConstructionService.Common/StateManager.cs]: There's this:

```
public T Property => CalledEverytime();
```
and
```
public T Property {get;} = CalledOnce();
```

**Review by premun — APPROVED** [review]

**Pull request merged**